File: src/2025-03-04-03-28-40--deepseek-chat-0304/python/exercises/practice/tree-building/tree_building.py
class Record:
    def __init__(self, record_id, parent_id):
        self.record_id = record_id
        self.parent_id = parent_id


class Node:
    def __init__(self, node_id):
        self.node_id = node_id
        self.children = []


def BuildTree(records):
    if not records:
        return None
        
    # Sort records by ID and validate continuity
    records.sort(key=lambda x: x.record_id)
    ordered_ids = [r.record_id for r in records]
    
    # Validate record IDs are continuous starting from 0
    if ordered_ids != list(range(len(records))):
        raise ValueError('Record id is invalid or out of order.')
    
    # Create nodes and build lookup dictionary
    nodes = {r.record_id: Node(r.record_id) for r in records}
    
    # Build tree structure
    for record in records:
        if record.record_id == 0:
            # Validate root node
            if record.parent_id != 0:
                raise ValueError("Node parent_id should be smaller than it's record_id.")
            continue
            
        # Validate parent-child relationship
        if record.record_id == record.parent_id:
            raise ValueError('Only root should have equal record and parent id.')
        if record.record_id < record.parent_id:
            raise ValueError("Node parent_id should be smaller than it's record_id.")
            
        # Get parent node and add current node as its child
        if record.parent_id not in nodes:
            raise ValueError('Parent node not found')
            
        parent_node = nodes[record.parent_id]
        parent_node.children.append(nodes[record.record_id])
    
    return nodes[0]

File: src/2025-03-04-03-28-40--deepseek-chat-0304/go/exercises/practice/tree-building/tree_building.go
package tree

import (
	"errors"
	"sort"
)

type Record struct {
	ID     int
	Parent int
}

type Node struct {
	ID       int
	Children []*Node
}

func Build(records []Record) (*Node, error) {
	if len(records) == 0 {
		return nil, nil
	}

	// Sort records by ID to ensure we process them in order
	sort.Slice(records, func(i, j int) bool {
		return records[i].ID < records[j].ID
	})

	// Validate records
	for i, r := range records {
		if r.ID != i {
			return nil, errors.New("records must be continuous")
		}
		if r.ID == 0 && r.Parent != 0 {
			return nil, errors.New("root node must have itself as parent")
		}
		if r.ID != 0 && r.Parent >= r.ID {
			return nil, errors.New("parent ID must be lower than child ID")
		}
	}

	// Create a map for quick node lookup
	nodes := make(map[int]*Node)
	for _, r := range records {
		nodes[r.ID] = &Node{ID: r.ID}
	}

	// Build the tree
	for _, r := range records {
		if r.ID == 0 {
			continue // Skip root node
		}

		parent := nodes[r.Parent]
		child := nodes[r.ID]
		parent.Children = append(parent.Children, child)
	}

	return nodes[0], nil
}
